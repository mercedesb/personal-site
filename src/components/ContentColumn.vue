<template>
  <div :class="'ContentColumn ContentColumn--' + color">
    <smart-link class='ContentColumn-link' :to="link" :isExternal="isExternal">
      <h2 class='ContentColumn-title'>{{ title }}</h2>
      <img class='ContentColumn-icon' :src="iconUrl" />
      <div class='ContentColumn-description'>
        <p>{{ preamble }}</p>
      </div>
    </smart-link>
  </div>
</template>

<script>
export default {
  props: {
    title: {
      type: String,
      required: true
    },
    color: {
      type: String,
      default: 'brown'
    },
    preamble: String,
    icon: Object,
    externalLink: String,
    urlSegment: String
  },
  computed: {
    iconUrl: function() {
      // return ''
      return this.getImageUrl(this.icon)
    },
    isExternal: function() {
      return !!this.externalLink
    },
    link: function() {
      if (this.isExternal) {
        return this.externalLink
      }
      else {
        return this.urlSegment
      }
    }
  }
}
</script>

<style lang="scss" scoped>
  $icon-width: 125px;
  $fixed-width: 340px;
  $transition-time: .5s;

  .ContentColumn {
    flex: 1 1 100%;
    text-align: center;
    padding-bottom: $large-spacing;
    -webkit-transition: flex-grow $transition-time;
    transition: flex-grow $transition-time;
    display: flex;
    flex-direction: column;
    align-items: center;

    &-link {
      font-weight: 400;
      width: $fixed-width;
    }

    &-title {
      padding-top: $large-spacing;
      color: $white;
    }

    &-icon {
      padding-top: $base-spacing;
      color: $white;
      width: $icon-width;
      z-index: 1;
    }

    &-description {
      color: $white;
      margin: 0 $base-spacing;
      font-weight: $base-font-weight;
    }

    @include media($min-tablet) {
      flex: 1 1 50%;
    }

    @include media($min-desktop) {
      flex: 1 1 $fixed-width;
      &:hover {
        flex-grow: 2;
      }
    }

    @include background-color;
    @include hover-darken;
  }
</style>