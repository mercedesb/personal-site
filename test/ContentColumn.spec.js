import { shallowMount, mount } from '@vue/test-utils'
import ContentColumn from '@/components/ContentColumn.vue'

describe('ContentColumn', () => {
  const initialProps = {
    title: 'title',
    color: 'blue',
    preamble: 'test preamble',
    icon: {
      fields: {
        file: {
          url: 'http://placeholder.pics/svg/300'
        }
      }
    },
    externalLink: '',
    urlSegment: 'testPath'
  }

  const shallow = propsData => shallowMount(ContentColumn, { 
    propsData: {
      ...initialProps,
      ...propsData
    }
   })

  const deep = propsData => mount(ContentColumn, {
    propsData: {
      ...initialProps,
      ...propsData
    }
  })


  describe('Snapshots', () => {
    let component 

    describe('with a color passed in', () => {
      it('matches snapshot', () => {
        component = shallow()
        expect(component.element).toMatchSnapshot()
      })
    })

    describe('without a color passed in', () => {
      it('matches snapshot', () => {
        component = shallow({color: ''})
        expect(component.element).toMatchSnapshot()
      })
    })
  })

  describe('Properties', () => {
    let component 

    beforeEach(() => {
      component = deep()
    })

    it('has a title property', () => {
      expect(component.props().title).toBe('title')
    })

    it('has a preamble property', () => {
      expect(component.props().preamble).toBe('test preamble')
    })

    it('has a color property', () => {
      expect(component.props().color).toBe('blue')
    })

    it('has a icon property', () => {
      expect(component.props().icon).toBeDefined()
    })

    it('has a urlSegment property', () => {
      expect(component.props().urlSegment).toBe('testPath')
    })
  })

  describe('Computed', () => {
    let component

    describe('isExternal', () => {
      it('returns true if externalLink prop is defined', () => {
        component = shallow({externalLink: 'http://google.com'})
        expect(component.vm.isExternal).toBe(true)
      })

      it('returns false if externalLink prop is not defined', () => {
        component = shallow()
        expect(component.vm.isExternal).toBe(false)
      })
    })

    describe('link', () => {
      let component

      it('returns externalLink from prop if defined', () => {
        component = shallow({externalLink: 'http://google.com'})
        expect(component.vm.link).toBe('http://google.com')
      })

      it('returns urlSegment if externalLink does not exist', () => {
        component = shallow()
        expect(component.vm.link).toBe('testPath')
      })
    })
  })
})