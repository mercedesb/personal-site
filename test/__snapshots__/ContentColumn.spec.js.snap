// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentColumn Snapshots with a color passed in matches snapshot 1`] = `
<div
  class="ContentColumn ContentColumn--blue"
>
  <smart-link
    class="ContentColumn-link"
    to="testPath"
  >
    <h2
      class="ContentColumn-title"
    >
      title
    </h2>
     
    <img
      class="ContentColumn-icon"
      src=""
    />
     
    <div
      class="ContentColumn-description"
    >
      <p>
        test preamble
      </p>
    </div>
  </smart-link>
</div>
`;

exports[`ContentColumn Snapshots without a color passed in matches snapshot 1`] = `
<div
  class="ContentColumn ContentColumn--"
>
  <smart-link
    class="ContentColumn-link"
    to="testPath"
  >
    <h2
      class="ContentColumn-title"
    >
      title
    </h2>
     
    <img
      class="ContentColumn-icon"
      src=""
    />
     
    <div
      class="ContentColumn-description"
    >
      <p>
        test preamble
      </p>
    </div>
  </smart-link>
</div>
`;
